CIP-GuardScripts/README.md

```diff
@@ -0,0 +1,138 @@
+---
+CIP: ????
+Title: GuardScriptCredential and Guard Script Purpose
+Category: Ledger
+Status: Proposed
+Authors:
+    - Philip DiSarro <philipdisarro@gmail.com>
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/???
+Created: 2025-07-24
+License: CC-BY-4.0
+---
+
+## Abstract
+
+This CIP proposes the introduction of a new script credential type, `GuardScriptCredential`, and a corresponding `Guard` script purpose. This enhancement allows a smart contract to validate not only UTxO spends from its address, but also UTxO creations to its address. The lack of such a mechanism today forces developers to implement complex workarounds involving authentication tokens, threaded NFTs, and registry UTxOs to guard against unauthorized or malformed deposits. This CIP aims to provide a native mechanism to guard script addresses against incoming UTxOs, thereby improving protocol safety, reducing engineering overhead, and eliminating a wide class of vulnerabilities in the Cardano smart contract ecosystem.
+
+## Motivation: why is this CIP necessary?
+
+In Cardano’s current eUTxO model, smart contracts can enforce logic only when their locked UTxOs are being spent. They have no ability to reject or validate UTxOs being sent to them. This leads to a fundamental weakness: anyone can send arbitrary tokens and datum to a script address, potentially polluting its state or spoofing valid contract UTxOs. To mitigate this, developers today must:
+
+- Mint authentication tokens
+- Use threading tokens to track contract state
+- Build registry systems with always-fails scripts
+- Validate datums defensively with token-datum context coupling
+
+These workarounds add significant complexity, on-chain cost, and surface area for bugs. A native mechanism to guard UTxO creations at a script address would eliminate the need for most of these patterns.
+
+## Specification
+We propose:
+
+1. **A new credential type**:
+   ```haskell
+   data Credential =
+       KeyCredential PubKeyHash
+     | ScriptCredential ScriptHash
+     | GuardScriptCredential ScriptHash -- new
+   ```
+2. **A new ScriptPurpose:**
+   ```haskell
+    data ScriptPurpose =
+        Spending TxOutRef
+    | Minting CurrencySymbol
+    | Certifying DCert
+    | Rewarding StakingCredential
+    | Voting Vote
+    | Proposing Proposal
+    | Guarding ScriptHash Integer (Maybe Integer) -- new
```

Comment:
Why do we need the `ScriptHash`? None of the other purposes includes `ScriptHash`.

Reply:
For completeness, can you include a short explanation of when and why `ScriptHash` is useful for Spending and Guarding?